### PR TITLE
Implement NVRTC interface

### PR DESCRIPTION
**Pull request by okuta**
This PR based on #33.

### CODE REVIEW

**Comment by delta2323** [comment]
Change `test_compiler.py` to change from `nvcc` to `nvrtc`.

**Comment by delta2323** [comment]
The description of `CUDA_PATH` in [`environment.rst`](https://github.com/cupy/cupy/blob/dfe94a2bb521e3028af7f9ca661c98cf02808442/docs/source/reference/environment.rst#id3) should be changed.

File: cupy/cuda/compiler.py [hunk 1]
```diff
@@ -2,35 +2,31 @@
...
-        _nvcc_version = _run_nvcc(cmd, '.')
-
-    return _nvcc_version
+    return _nvrtc_version
 
 
 def _get_arch():
     cc = device.Device().compute_capability
-    return 'sm_%s' % cc
+    return 'compute_%s' % cc
```
Comment (mitmul) [inline cupy/cuda/compiler.py]: Sorry for inconvenience, I replied to it in my comment at this PR:
https://github.com/cupy/cupy/pull/33#discussion_r113092304

> compute_xx is used to specify a virtual GPU architecture, while sm_xx is used to specify the actual real architecture for which a given code/PTX will be compiled. NVRTC generates PTX file for virtual architecture, and then the CUDA driver converts the PTX file for each real GPU architecture by generating an optimized binary.

So in this case, we need to use compute_xx for the option to NVRTC.

File: tests/cupy_tests/linalg_tests/test_product.py [hunk 2]
```diff
@@ -259,34 +261,58 @@ def test_transposed_tensordot(self, xp, dtype):
     @testing.for_all_dtypes()
     @testing.numpy_cupy_allclose()
     def test_tensordot_with_int_axes(self, xp, dtype):
-        a = testing.shaped_arange((2, 3, 4, 5), xp, dtype)
-        b = testing.shaped_arange((3, 4, 5, 2), xp, dtype)
-        return xp.tensordot(a, b, axes=3)
+        if dtype in (numpy.uint8, numpy.int8, numpy.uint16, numpy.int16):
```
Comment (okuta) [inline tests/cupy_tests/linalg_tests/test_product.py]: For avoid overflow.

Comment (delta2323) [inline tests/cupy_tests/linalg_tests/test_product.py]: OK

File: tests/cupy_tests/core_tests/test_fusion.py [hunk 3]
```diff
@@ -887,7 +887,7 @@ def test_fuse5(self, xp, dtype):
         a = xp.arange(15, dtype=dtype)
         b = a * a[::-1]
         a = a * 3 + 11
-        c = (a * b) ** 2 % 63
+        c = (a * b) % 63
```
Comment (okuta) [inline tests/cupy_tests/core_tests/test_fusion.py]: This is also avoiding overflow.

File: cupy/cuda/compiler.py (hunk 1 above)
Comment (delta2323) [inline cupy/cuda/compiler.py]: OK, thanks.

File: cupy/cuda/cupy_cuda.h [hunk 4]
```diff
@@ -59,18 +59,23 @@ typedef int CUdevice;
...
 struct CUmod_st;
 struct CUstream_st;
+struct CUlinkState_st;
+
 
 typedef struct CUevent_st* cudaEvent_t;
 typedef struct CUfunc_st* CUfunction;
 typedef struct CUmod_st* CUmodule;
 typedef struct CUstream_st* cudaStream_t;
+typedef struct CUlinkState_st *CUlinkState;
```
Comment (delta2323) [inline cupy/cuda/cupy_cuda.h]: Should it be `typedef struct CUlinkState_st* CUlinkState`?

File: cupy/cuda/cupy_nvrtc.h [hunk 5]
```diff
@@ -0,0 +1,60 @@
+// This file is a stub header file of cudnn for Read the Docs.
```
Comment (delta2323) [inline cupy/cuda/cupy_nvrtc.h]: cudnn -> nvrtc

**Comment by delta2323** [comment]
Jenkins test failed could you check that? All of Py2, Py3, Py35 configurations failed. But I want you to check only Py3 and Py35 because the the CUDA version of Py2 is 6.0, which is out of scope of this PR.

File: tests/cupy_tests/core_tests/test_fusion.py (hunk 3 above)
Comment (delta2323) [inline tests/cupy_tests/core_tests/test_fusion.py]: ok

**Comment by okuta** [comment]
@delta2323 I fixed.

**Comment by delta2323** [comment]
Could you check the result of Jenkins (cupy_master_chainer#235 )?

**Comment by okuta** [comment]
@delta2323 I updated test. Please review this.

**Comment by okuta** [comment]
Please merge chainer/chainer#2877 before this PR.

**Comment by okuta** [comment]
@delta2323 Please merge this PR.

**Comment by delta2323** [comment]
I launched GPU tests of CuPy and Chainer.

**Comment by delta2323** [comment]
GPU tests passed.

**Pull request merged**